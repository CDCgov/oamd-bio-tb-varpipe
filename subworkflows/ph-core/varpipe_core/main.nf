//
// varpipe core subworkflow
//

include { TRIMMOMATIC                                                           } from '../../../modules/ph-core/trimmomatic/main'
include { VARPIPE_PROCESSBAM                                                    } from '../varpipe_processbam/main'
include { VARPIPE_COVERAGEANALYSIS                                              } from '../../../modules/ph-core/varpipe/coverageanalysis/main'
include { VARPIPE_STRUCTURALVARIANTS                                            } from '../../../modules/ph-core/varpipe/structuralvariants/main'
include { VARPIPE_TARGETEDVARIANTS                                              } from '../varpipe_targetedvariants/main'
include { VARPIPE_ALLVARIANTS                                                   } from '../varpipe_allvariants/main'
include { VARPIPE_CREATEANNOTATIONS as VARPIPE_CREATEANNOTATIONS_TARGETED       } from '../../../modules/ph-core/varpipe/createannotations/main'
include { VARPIPE_CREATEANNOTATIONS as VARPIPE_CREATEANNOTATIONS_ALL            } from '../../../modules/ph-core/varpipe/createannotations/main'
include { VARPIPE_PARSEANNOTATIONS as VARPIPE_PARSEANNOTATIONS_TARGETED         } from '../../../modules/ph-core/varpipe/parseannotations/main'
include { VARPIPE_PARSEANNOTATIONS as VARPIPE_PARSEANNOTATIONS_ALL              } from '../../../modules/ph-core/varpipe/parseannotations/main'
include { VARPIPE_LINEAGE                                                       } from '../../../modules/ph-core/varpipe/lineage/main'
include { VARPIPE_REPORT                                                        } from '../../../modules/ph-core/varpipe/report/main'
include { VARPIPE_INTERPRETATION                                                } from '../../../modules/ph-core/varpipe/interpretation/main'
include { VARPIPE_PDF                                                           } from '../../../modules/ph-core/varpipe/pdf/main'
include { VARPIPE_TAR                                                           } from '../../../modules/ph-core/varpipe/tar/main'

workflow VARPIPE_CORE {
    take:
    ch_cleaned_reads // channel (mandatory): [ val(meta), path(reads) ]
    datestamp        // string (mandatory): timestamp for output identification
    ref_fasta        // channel (mandatory): [ path(params.ref_fasta) ]
    bedlist_amp      // channel (mandatory): [ path(params.bedlist_amp) ]
    bedlist_one      // channel (mandatory): [ path(params.bedlist_one) ]
    bedlist_two      // channel (mandatory): [ path(params.bedlist_two) ]
    bedstruct        // channel (mandatory): [ path(params.bedstruct) ]
    mutationloci     // channel (mandatory): [ path(params.mutationloci) ]
    intervals        // channel (mandatory): [ path(params.intervals) ]
    reported         // channel (mandatory): [ path(params.reported) ]
    lineages         // channel (mandatory): [ path(params.lineages) ]
    snpeff_db        // channel (mandatory): [ path(params.snpeff_db) ]

    main:
    ch_versions = Channel.empty()

    // QC TRIMMOMATIC
    TRIMMOMATIC(
        ch_cleaned_reads
    )
    ch_versions = ch_versions.mix(TRIMMOMATIC.out.versions)

    ch_ref_fasta = Channel.fromPath(ref_fasta)
        .map { fasta -> [ [ id: 'reference' ], fasta] }
        .collect()

    VARPIPE_PROCESSBAM (
        TRIMMOMATIC.out.trimmed_reads,
        ch_ref_fasta
    )
    ch_versions = ch_versions.mix(VARPIPE_PROCESSBAM.out.versions)

    ch_coverageanalysis_input = VARPIPE_PROCESSBAM.out.coverage
        .join(VARPIPE_PROCESSBAM.out.unmapped)
        .join(VARPIPE_PROCESSBAM.out.mapped)

    VARPIPE_COVERAGEANALYSIS(
        ch_coverageanalysis_input,
        bedlist_amp,
        [bedlist_one, bedlist_two],
    )
    ch_versions = ch_versions.mix(VARPIPE_COVERAGEANALYSIS.out.versions)

    VARPIPE_STRUCTURALVARIANTS(
        VARPIPE_COVERAGEANALYSIS.out.qc_passed.join(ch_coverageanalysis_input),
        VARPIPE_COVERAGEANALYSIS.out.genome_region_coverage,
        bedstruct
    )
    ch_versions = ch_versions.mix(VARPIPE_STRUCTURALVARIANTS.out.versions)

    // CALLERS
    ch_gatk_targeted_input = VARPIPE_COVERAGEANALYSIS.out.qc_passed
        .map { meta1, _qc -> [meta1] }
        .join(VARPIPE_PROCESSBAM.out.bam)
        .join(VARPIPE_PROCESSBAM.out.bai)
        .map { meta1, bam, bai ->
            return [meta1, bam, bai, intervals]
        }

    VARPIPE_TARGETEDVARIANTS(
        ch_gatk_targeted_input,
        VARPIPE_PROCESSBAM.out.fai.collect(),
        VARPIPE_PROCESSBAM.out.reference_dict.collect(),
        ch_ref_fasta,
        snpeff_db
    )
    ch_versions = ch_versions.mix(VARPIPE_TARGETEDVARIANTS.out.versions)

    ch_gatk_whole_genome_input = VARPIPE_COVERAGEANALYSIS.out.qc_passed
        .map { meta1, _qc -> [meta1] }
        .join(VARPIPE_PROCESSBAM.out.bam)
        .join(VARPIPE_PROCESSBAM.out.bai)
        .map { meta1, bam, bai ->
            return [meta1, bam, bai, []]
        }

    VARPIPE_ALLVARIANTS(
        ch_gatk_whole_genome_input,
        VARPIPE_PROCESSBAM.out.fai.collect(),
        VARPIPE_PROCESSBAM.out.reference_dict.collect(),
        ch_ref_fasta,
        snpeff_db
    )
    ch_versions = ch_versions.mix(VARPIPE_ALLVARIANTS.out.versions)

    VARPIPE_CREATEANNOTATIONS_TARGETED(
        VARPIPE_TARGETEDVARIANTS.out.ann
    )
    ch_versions = ch_versions.mix(VARPIPE_CREATEANNOTATIONS_TARGETED.out.versions.first())

    VARPIPE_CREATEANNOTATIONS_ALL(
        VARPIPE_ALLVARIANTS.out.ann
    )
    ch_versions = ch_versions.mix(VARPIPE_CREATEANNOTATIONS_ALL.out.versions.first())


    VARPIPE_PARSEANNOTATIONS_TARGETED(
        VARPIPE_CREATEANNOTATIONS_TARGETED.out.drLociAnnotation,
        mutationloci
    )
    ch_versions = ch_versions.mix(VARPIPE_PARSEANNOTATIONS_TARGETED.out.versions.first())

    VARPIPE_PARSEANNOTATIONS_ALL(
        VARPIPE_CREATEANNOTATIONS_ALL.out.fullAnnotation,
        mutationloci
    )
    ch_versions = ch_versions.mix(VARPIPE_PARSEANNOTATIONS_ALL.out.versions.first())

    // RUN VARPIPE_LINEAGE ANALYSIS
    VARPIPE_LINEAGE(
        VARPIPE_PARSEANNOTATIONS_ALL.out.fullFinalAnnotation,
        lineages
    )
    ch_versions = ch_versions.mix(VARPIPE_LINEAGE.out.versions.first())

    VARPIPE_COVERAGEANALYSIS.out.stats
        .join(VARPIPE_COVERAGEANALYSIS.out.target_region_coverage)
        .join(VARPIPE_PARSEANNOTATIONS_TARGETED.out.drLociFinalAnnotation)
        .set { ch_varpipe_report }

    // GENERATE ANALYSIS REPORT, INTERPRETATION & PDF REPORT

    VARPIPE_REPORT(
        ch_varpipe_report
    )
    ch_versions = ch_versions.mix(VARPIPE_REPORT.out.versions.first())

    VARPIPE_REPORT.out.summary
        .join(VARPIPE_STRUCTURALVARIANTS.out.structural_variants)
        .join(VARPIPE_CREATEANNOTATIONS_TARGETED.out.drLociAnnotation)
        .join(VARPIPE_COVERAGEANALYSIS.out.target_region_coverage)
        .set { ch_varpipe_interpretation }

    VARPIPE_INTERPRETATION(
        ch_varpipe_interpretation,
        reported
    )
    ch_versions = ch_versions.mix(VARPIPE_INTERPRETATION.out.versions.first())

    VARPIPE_PDF(
        VARPIPE_INTERPRETATION.out.summary
    )
    ch_versions = ch_versions.mix(VARPIPE_PDF.out.versions.first())

    ch_varpipe_outputs = Channel.empty()
        .mix(
            VARPIPE_COVERAGEANALYSIS.out.stats,
            VARPIPE_COVERAGEANALYSIS.out.qc_failed,
            VARPIPE_COVERAGEANALYSIS.out.target_region_coverage,
            VARPIPE_COVERAGEANALYSIS.out.genome_region_coverage,
            VARPIPE_STRUCTURALVARIANTS.out.structural_variants,
            VARPIPE_LINEAGE.out.lineage,
            VARPIPE_LINEAGE.out.lineage_report,
            VARPIPE_INTERPRETATION.out.summary,
            VARPIPE_INTERPRETATION.out.interpretation,
            VARPIPE_PDF.out.report,
            VARPIPE_PARSEANNOTATIONS_TARGETED.out.drLociFinalAnnotation,
            VARPIPE_CREATEANNOTATIONS_TARGETED.out.drLociAnnotation,
            VARPIPE_TARGETEDVARIANTS.out.ann,
            VARPIPE_ALLVARIANTS.out.ann,
            VARPIPE_PARSEANNOTATIONS_ALL.out.fullFinalAnnotation,
            VARPIPE_CREATEANNOTATIONS_ALL.out.fullAnnotation
        )
        .groupTuple()
        .multiMap { meta, files ->
            meta: [ id: "${meta.id}", filenames: files.collect { it.getName() } ]
            files: files
        }

    VARPIPE_TAR(
        ch_varpipe_outputs.meta.collect(),
        ch_varpipe_outputs.files.collect(),
        datestamp
    )
    ch_versions = ch_versions.mix(VARPIPE_TAR.out.versions)

    emit:
    lineage = VARPIPE_LINEAGE.out.lineage_report // channel (mandatory): [ val(meta), path(lineage_report) ]
    versions = ch_versions // channel (mandatory): [ versions.yml ]
}
