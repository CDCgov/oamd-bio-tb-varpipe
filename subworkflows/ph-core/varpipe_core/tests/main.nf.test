nextflow_workflow {

    name "Test Subworkflow VARPIPE_CORE"
    script "../main.nf"
    workflow "VARPIPE_CORE"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_phcore"
    tag "subworkflows/varpipe_core"
    tag "subworkflows/varpipe_processbam"
    tag "subworkflows/varpipe_targetedvariants"
    tag "subworkflows/varpipe_allvariants"
    tag "cat/fastq"
    tag "trimmomatic"
    tag "picard/buildbamindex"
    tag "picard/markduplicates"
    tag "picard/samformatconverter"
    tag "picard/sortsam"
    tag "picard/createsequencedictionary"
    tag "samtools/faidx"
    tag "samtools/depth"
    tag "samtools/view"
    tag "bwa/index"
    tag "varpipe/bwamem"
    tag "gatk4/leftalignandtrimvariants"
    tag "gatk4/mutect2"
    tag "gatk4/filtermutectcalls"
    tag "varpipe/snpeff"
    tag "varpipe/parseannotations"
    tag "varpipe/createannotations"
    tag "varpipe/coverageanalysis"
    tag "varpipe/structuralvariants"
    tag "varpipe/report"
    tag "varpipe/interpretation"
    tag "varpipe/lineage"
    tag "varpipe/pdf"
    tag "varpipe/tar"

    setup {
        run("CAT_FASTQ") {
            script "modules/ph-core/cat/fastq/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [
                    file(params.modules_testdata_base_path + "/genomics/tuberculosis/fastqs/SRR24827267_1.part_001.fastq.gz"),
                    file(params.modules_testdata_base_path + "/genomics/tuberculosis/fastqs/SRR24827267_2.part_001.fastq.gz"),
                    file(params.modules_testdata_base_path + "/genomics/tuberculosis/fastqs/SRR24827267_1.part_002.fastq.gz"),
                    file(params.modules_testdata_base_path + "/genomics/tuberculosis/fastqs/SRR24827267_2.part_002.fastq.gz"),
                    file(params.modules_testdata_base_path + "/genomics/tuberculosis/fastqs/SRR24827267_1.part_003.fastq.gz"),
                    file(params.modules_testdata_base_path + "/genomics/tuberculosis/fastqs/SRR24827267_2.part_003.fastq.gz"),
                    file(params.modules_testdata_base_path + "/genomics/tuberculosis/fastqs/SRR24827267_1.part_004.fastq.gz"),
                    file(params.modules_testdata_base_path + "/genomics/tuberculosis/fastqs/SRR24827267_2.part_004.fastq.gz")
                    ]
                ])
                """
            }
        }
    }

    test("varpipe-core subworkflow test [paired-end]") {

        when {
            params {
                date = "TEST_DATESTAMP"
                keepfiles = false\
            }
            workflow {
                """
                input[0] = CAT_FASTQ.out.reads
                input[1] = params.date
                input[2] = file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/kb/varpipe/ref2.fa')
                input[3] = file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/kb/varpipe/amp_bed.txt')
                input[4] = file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/kb/varpipe/bed_1.txt')
                input[5] = file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/kb/varpipe/bed_2.txt')
                input[6] = file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/kb/varpipe/BED.txt')
                input[7] = file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/kb/varpipe/mutation_loci.txt')
                input[8] = file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/kb/varpipe/intervals.bed')
                input[9] = file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/kb/varpipe/reportable.txt')
                input[10] = file(params.modules_testdata_base_path + '/genomics/tuberculosis/genome/kb/varpipe/lineage_markers.txt')
                input[11] = 'NC_000962'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
