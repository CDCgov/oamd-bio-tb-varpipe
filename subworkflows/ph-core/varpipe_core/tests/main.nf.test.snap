{
    "varpipe-core subworkflow test [paired-end]": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.lineage_report.txt:md5,33de4c0178efcfe32884cba8b5884b99"
                    ]
                ],
                "1": [
                    "versions.yml:md5,083f9179e174ad4224d0f983226d27cf",
                    "versions.yml:md5,0f2251602a6f367cca44ce4ac5ac37cd",
                    "versions.yml:md5,0fc5f92564fed61db8838c799288f853",
                    "versions.yml:md5,18d71031bd7c4f21bae23f5c1fb79ddd",
                    "versions.yml:md5,1fd304adc95521c9bee131237688b22a",
                    "versions.yml:md5,2a7e2f1529a97e493b756fd9e0171622",
                    "versions.yml:md5,32bd798da61d5f94a10fc02b761991ce",
                    "versions.yml:md5,400f606a4c3d261af2a1d4ec850d7d98",
                    "versions.yml:md5,4421f6a603f88d7003e78fe54100c574",
                    "versions.yml:md5,4841885687a2923c06bb18c0fdec2593",
                    "versions.yml:md5,533207d6296de6e78e7b9b7763efa05d",
                    "versions.yml:md5,5b6d121f3123b87fcaf7a9f3bb857cc5",
                    "versions.yml:md5,63c7dccf550f918ea2fcb5998a420e00",
                    "versions.yml:md5,73c6a752e5f8e31a495e72092e0d3562",
                    "versions.yml:md5,79d6c180441246da2c8cc6a3ffa5046c",
                    "versions.yml:md5,807e3ea08f79a4ffcc9f5ba66e11dc6b",
                    "versions.yml:md5,81d1a2e61a4ea58f65a055d815be3bbd",
                    "versions.yml:md5,8656a2312fc4fb2cd24ceaace578b8d3",
                    "versions.yml:md5,899949bc4665b4d0698b6e8a068d48d1",
                    "versions.yml:md5,8a7f7b6ecb251a73f9472a5d2ee446df",
                    "versions.yml:md5,8cf5fc28cd88a175bef6a009dcab96c3",
                    "versions.yml:md5,8e8c99cfbab9c085f6786709436fb218",
                    "versions.yml:md5,905e95b2e3b27500470cfa3759d8f486",
                    "versions.yml:md5,a27549eb7ad5184443634a28c3175965",
                    "versions.yml:md5,b01f1895b1b41da5cea14a55efe8b80c",
                    "versions.yml:md5,b22d24a19a76807dd5f731c2ce75ac2c",
                    "versions.yml:md5,b78ab17c4489c3113145eed954eed1c6",
                    "versions.yml:md5,bbe4172e9805d98a6b7a209940796d13",
                    "versions.yml:md5,dba1844e65849f85bd3384986ff6e4f8",
                    "versions.yml:md5,e2b79d9a1e7daf2ab0496fef48b40794",
                    "versions.yml:md5,e342b7df18490919d1af0a1cd68e9fca",
                    "versions.yml:md5,f393d791bb0a72a5bfd1c591876024b7",
                    "versions.yml:md5,fbbc2fcc2af61637a2488020a7de6435"
                ],
                "lineage": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.lineage_report.txt:md5,33de4c0178efcfe32884cba8b5884b99"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,083f9179e174ad4224d0f983226d27cf",
                    "versions.yml:md5,0f2251602a6f367cca44ce4ac5ac37cd",
                    "versions.yml:md5,0fc5f92564fed61db8838c799288f853",
                    "versions.yml:md5,18d71031bd7c4f21bae23f5c1fb79ddd",
                    "versions.yml:md5,1fd304adc95521c9bee131237688b22a",
                    "versions.yml:md5,2a7e2f1529a97e493b756fd9e0171622",
                    "versions.yml:md5,32bd798da61d5f94a10fc02b761991ce",
                    "versions.yml:md5,400f606a4c3d261af2a1d4ec850d7d98",
                    "versions.yml:md5,4421f6a603f88d7003e78fe54100c574",
                    "versions.yml:md5,4841885687a2923c06bb18c0fdec2593",
                    "versions.yml:md5,533207d6296de6e78e7b9b7763efa05d",
                    "versions.yml:md5,5b6d121f3123b87fcaf7a9f3bb857cc5",
                    "versions.yml:md5,63c7dccf550f918ea2fcb5998a420e00",
                    "versions.yml:md5,73c6a752e5f8e31a495e72092e0d3562",
                    "versions.yml:md5,79d6c180441246da2c8cc6a3ffa5046c",
                    "versions.yml:md5,807e3ea08f79a4ffcc9f5ba66e11dc6b",
                    "versions.yml:md5,81d1a2e61a4ea58f65a055d815be3bbd",
                    "versions.yml:md5,8656a2312fc4fb2cd24ceaace578b8d3",
                    "versions.yml:md5,899949bc4665b4d0698b6e8a068d48d1",
                    "versions.yml:md5,8a7f7b6ecb251a73f9472a5d2ee446df",
                    "versions.yml:md5,8cf5fc28cd88a175bef6a009dcab96c3",
                    "versions.yml:md5,8e8c99cfbab9c085f6786709436fb218",
                    "versions.yml:md5,905e95b2e3b27500470cfa3759d8f486",
                    "versions.yml:md5,a27549eb7ad5184443634a28c3175965",
                    "versions.yml:md5,b01f1895b1b41da5cea14a55efe8b80c",
                    "versions.yml:md5,b22d24a19a76807dd5f731c2ce75ac2c",
                    "versions.yml:md5,b78ab17c4489c3113145eed954eed1c6",
                    "versions.yml:md5,bbe4172e9805d98a6b7a209940796d13",
                    "versions.yml:md5,dba1844e65849f85bd3384986ff6e4f8",
                    "versions.yml:md5,e2b79d9a1e7daf2ab0496fef48b40794",
                    "versions.yml:md5,e342b7df18490919d1af0a1cd68e9fca",
                    "versions.yml:md5,f393d791bb0a72a5bfd1c591876024b7",
                    "versions.yml:md5,fbbc2fcc2af61637a2488020a7de6435"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-05-16T13:43:24.898424354"
    }
}
